Improvements to ContextualService

lucko opened the pull request:
Fixes #1546 - still WIP for now

The contexts system in particular can be difficult for developers new to Sponge (esp. sponge permissions) to understand - the hope is that some better javadocs will help a bit with that.


If anyone has anything they think could be clarified/improved/explained better please say. :)

lucko committed:
Improvements to ContextualService

Aaron1011 commented:
@lucko Are you planning to make any breaking changes? If not, this looks like like it could target `stable-7`

lucko commented:
Not sure yet, there *probably* won't be any, but I thought it made sense to target bleeding for now - can always cherry pick it back.

The only thing I've considered so far is changing ContextCalculator's

```java
void accumulateContexts(T target, Set<Context> accumulator)
```

to 

```java
void accumulateContexts(T target, Consumer<Context> accumulator)
```

reasoning being that calculators shouldn't be relying on the state of the Set, and should only be using `add` and `addAll` - not sure it's worth the break though 🤷‍♂️

lucko commented:
This is ready now.

* Mostly changes to javadocs
* No breaking changes, only one new static method added, so this can be backported to stable-7
* Two methods in `Context` were deprecated - reasoning explained in the doc. They can be removed in the next API release.


@bloodmc @Aaron1011 or anyone else kind enough to, would you mind giving the doc changes a quick read over to make sure they make sense. I'm finding it quite hard to put a lot of the concepts into words - it's almost too abstract in places to explain. :)

gabizou commented:
Want to update the PR to target `stable-7` maybe?

ST-DDT commented on src/main/java/org/spongepowered/api/service/context/ContextCalculator.java:
```diff
@@ -60,5 +141,9 @@
      *         false.
      */
-    boolean matches(Context context, T calculable);
+    default boolean matches(Context context, T calculable) {
```
I would rename `calculable` to `target` to match the other method

lucko replied:
Done, thanks

lucko commented:
Updated (not so successfully, sorry for the review request pings 😢) to target `stable-7`.

Also addressed ST-DDT's suggestion. This is ready to be merged now, unless anyone has any final suggestions / comments.
